File: Sqrt(x).py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x == 0:
            return 0

        # Binary search based on the condition: 
        # if (mid * mid < x and (mid + 1) * (mid + 1) > x) or mid * mid == x, then the answer is mid

        left = 1
        right = x

        while left <= right:
            mid = (left + right) // 2

            if mid * mid > x: #mid is too big, need to check the left side
                right = mid - 1
            elif (mid * mid < x and (mid + 1) * (mid + 1) > x) or mid * mid == x:
                return mid
            else:
                left = mid + 1

        #Time: O(log(x))
        #Space: O(1)

File: Sqrt(x).java
class Solution {
    public int mySqrt(int x) {
        if (x == 0){
            return 0;
        }

        int l = 1;
        int r = x;
        int mid = 1;
        
        while (l <= r){
            mid = l + ((r - l) / 2); //to avoid integer overflow

            if (mid * mid == x){
                return mid;
            }
            else if((long) mid * mid > (long) x){ //Add long as the number might be too big
                r = mid - 1;
            }
            else{
                l = mid + 1;
            }
        }
        //Once l > r, without returning anything inside the loop, the answer is r
        return r;
    }
}
// Time: O(log(n))
// Space: O(1)
